bftengine/src/bftengine/FullNodeReplica.cpp

```diff
@@ -0,0 +1,295 @@
+// Concord
+//
+// Copyright (c) 2018, 2019 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").  You may not use this product except in
+// compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright notices and license terms. Your use of
+// these subcomponents is subject to the terms and conditions of the sub-component's license, as noted in the LICENSE
+// file.
+
+#include <optional>
+#include <functional>
+#include <bitset>
+
+#include <bftengine/Replica.hpp>
+#include <messages/StateTransferMsg.hpp>
+#include "FullNodeReplica.hpp"
+
+#include "log/logger.hpp"
+#include "MsgHandlersRegistrator.hpp"
+#include "messages/CheckpointMsg.hpp"
+#include "messages/AskForCheckpointMsg.hpp"
+#include "messages/ClientRequestMsg.hpp"
+#include "messages/ClientReplyMsg.hpp"
+#include "util/kvstream.h"
+#include "PersistentStorage.hpp"
+#include "MsgsCommunicator.hpp"
+#include "SigManager.hpp"
+#include "ReconfigurationCmd.hpp"
+#include "util/json_output.hpp"
+#include "SharedTypes.hpp"
+#include "communication/StateControl.hpp"
+
+using concordUtil::Timers;
+
+namespace bftEngine::impl {
+
+FullNodeReplica::FullNodeReplica(const ReplicaConfig &config,
+                                 std::shared_ptr<IRequestsHandler> requestsHandler,
+                                 IStateTransfer *stateTransfer,
+                                 std::shared_ptr<MsgsCommunicator> msgComm,
+                                 std::shared_ptr<MsgHandlersRegistrator> msgHandlerReg,
+                                 concordUtil::Timers &timers,
+                                 MetadataStorage *metadataStorage)
+    : ReplicaForStateTransfer(config, requestsHandler, stateTransfer, msgComm, msgHandlerReg, true, timers),
+      fn_metrics_{metrics_.RegisterCounter("receivedCheckpointMsgs"),
+                  metrics_.RegisterCounter("sentAskForCheckpointMsgs"),
+                  metrics_.RegisterCounter("receivedInvalidMsgs"),
+                  metrics_.RegisterGauge("lastExecutedSeqNum", lastExecutedSeqNum)},
+      metadataStorage_{metadataStorage} {
+  LOG_INFO(GL, "Initialising Full Node Replica");
+  repsInfo = new ReplicasInfo(config, dynamicCollectorForPartialProofs, dynamicCollectorForExecutionProofs);
+  msgHandlers_->registerMsgHandler(
```

Comment:
nit: 
1)consider moving the 4 registeration functions to a seperate function
registerMsgHandlers
2) add `using namespace std::placeholders;` and shorten code.